libraries/ms-helper/src/can_uart.c

```diff
@@ -0,0 +1,106 @@
+#include "can_uart.h"
+#include <string.h>
+#include "log.h"
+
+// CTX
+#define CAN_UART_TX_MARKER 0x585443
+// CRX
+#define CAN_UART_RX_MARKER 0x585243
+
+// Defined protocol: Header (u32), ID (u32), Data (u64), newline (u8)
+// Bits  | Field
+// ------|--------------------
+// 0:23  | marker (CTX || CRX)
+// 24    | extended
+// 25    | rtr
+// 26-27 | reserved
+// 28:31 | dlc
+
+#define CAN_UART_BUILD_HEADER(marker, extended, rtr, dlc)             \
+  (((uint32_t)(marker)&0xFFFFFF) | ((uint32_t)(extended)&0x1) << 24 | \
+   ((uint32_t)(rtr)&0x1) << 25 | ((uint32_t)(dlc)&0xF) << 28)
+#define CAN_UART_EXTRACT_HEADER(header, marker, extended, rtr, dlc) \
+  ({                                                                \
+    uint32_t _header = header;                                      \
+    *(marker) = _header & 0xFFFFFF;                                 \
+    *(extended) = (_header >> 25) & 0x1;                            \
+    *(rtr) = (_header >> 25) & 0x1;                                 \
```

Comment:
Should this be the same as `extended`? You have the same shift

```c
*(extended) = (_header >> 25) & 0x1;
*(rtr) = (_header >> 25) & 0x1;
```

Reply:
whoops